request_handler_proxy.cpp

```diff
@@ -0,0 +1,187 @@
+	  	path = request.uri().substr(uri_prefix.length());
+	  	if (path.length() == 0)
+	  		path = "/";
+	  	std::cout << path << std::endl;
+	  	bool is_success = make_request(host, port, path, false, response);
+
+	  	if (!is_success) {
+	  		std::string fail_body = "Request to remote server failed";
+	  		response->SetStatus(Response::ResponseCode::NOT_FOUND);
+		  	response->AddHeader("Content-Type", "text/plain");
+		  	response->AddHeader("Content-Length", std::to_string(fail_body.length()));
+		  	response->SetBody(fail_body);
+	  	}
+
+		return RequestHandler::Status::OK;
+	}
+
+	bool ProxyHandler::make_request(std::string host, std::string port, std::string path, bool redirect, Response* res) {
+		std::cout << "Making request to: www." << host << ":" << port << path << std::endl;
+
+		using boost::asio::ip::tcp;
+		boost::asio::io_service io_service;
+
+
+	    // Get a list of endpoints corresponding to the server name.
+	    tcp::resolver resolver(io_service);
+	    tcp::resolver::query query(host, port);
+	    tcp::resolver::iterator endpoint_iterator = resolver.resolve(query);
+
+	    // Try each endpoint until we successfully establish a connection.
+	    tcp::socket socket(io_service);
+	    boost::asio::connect(socket, endpoint_iterator);
+
+	    // Form the request. We specify the "Connection: close" header so that the
+	    // server will close the socket after transmitting the response. This will
+	    // allow us to treat all data up until the EOF as the content.
+	    boost::asio::streambuf request;
+	    std::ostream request_stream(&request);
+	    request_stream << "GET " << path << " HTTP/1.0\r\n";
+	    if (!redirect)
+	    	request_stream << "Host: www." << host << ":" << port << "\r\n";
+	    else
+	    	request_stream << "Host: " << host << "\r\n";
+	    request_stream << "Accept: */*\r\n";
+	    request_stream << "Connection: close\r\n\r\n";
+
+	    // For debugging: print out the request
+	    boost::asio::streambuf::const_buffers_type bufs = request.data();
+		std::string str(boost::asio::buffers_begin(bufs), boost::asio::buffers_begin(bufs) + request.size());
+		std::cout << str << std::endl;
+
+	    // Send the request.
+	    boost::asio::write(socket, request);
+
+	    // Read the response status line. The response streambuf will automatically
+	    // grow to accommodate the entire line. The growth may be limited by passing
+	    // a maximum size to the streambuf constructor.
+	    boost::asio::streambuf response_buffer;
+	    boost::asio::read_until(socket, response_buffer, "\r\n");
+
+	    // Check that response is OK.
+	    std::istream response_stream(&response_buffer);
+
+	    std::string http_version;
+	    response_stream >> http_version;
+
+	    unsigned int status_code;
+	    response_stream >> status_code;
+
+	    std::string status_message;
+	    std::getline(response_stream, status_message);
+	    if (!response_stream || http_version.substr(0, 5) != "HTTP/")
```

Comment:
Nice way to check for HTTP and prevents a blank page to appear when you into a https website